creusot-contracts/src/std/pointer.rs

```diff
@@ -0,0 +1,92 @@
+use crate::*;
+
+// We conservatively model raw pointers as having an address *plus some hidden
+// metadata*.
+//
+// This is to account for provenance
+// (https://doc.rust-lang.org/std/ptr/index.html#using-strict-provenance) and
+// wide pointers. See e.g.
+// https://doc.rust-lang.org/std/primitive.pointer.html#method.is_null: "unsized
+// types have many possible null pointers, as only the raw data pointer is
+// considered, not their length, vtable, etc. Therefore, two pointers that are
+// null may still not compare equal to each other."
+
+pub trait PointerExt<T: ?Sized>: Sized {
+    #[logic]
+    #[ensures(result.addr_logic() == 0)]
+    fn null_logic() -> Self;
+
+    #[logic]
+    fn addr_logic(self) -> Int;
+}
+
+impl<T: ?Sized> PointerExt<T> for *const T {
+    #[trusted]
+    #[open(self)]
+    #[logic]
+    #[ensures(result.addr_logic() == 0)]
+    fn null_logic() -> Self {
+        absurd
+    }
+
+    #[trusted]
+    #[predicate]
+    #[open(self)]
+    fn addr_logic(self) -> Int {
+        absurd
+    }
+}
+
+impl<T: ?Sized> PointerExt<T> for *mut T {
+    #[trusted]
+    #[open(self)]
+    #[logic]
+    #[ensures(result.addr_logic() == 0)]
+    fn null_logic() -> Self {
+        absurd
+    }
+
+    #[trusted]
+    #[predicate]
+    #[open(self)]
+    fn addr_logic(self) -> Int {
+        absurd
+    }
+}
+
+extern_spec! {
+    impl<T> *const T {
+        #[ensures(result@ == self.addr_logic())]
+        fn addr(self) -> usize;
+
+        #[ensures(result == (self.addr_logic() == 0))]
+        fn is_null(self) -> bool;
+    }
+
+    impl<T> *mut T {
+        #[ensures(result@ == self.addr_logic())]
+        fn addr(self) -> usize;
+
+        #[ensures(result == (self.addr_logic() == 0))]
+        fn is_null(self) -> bool;
+    }
+
+    mod std {
+        mod ptr {
+            #[ensures(result.addr_logic() == 0)]
```

Comment:
```suggestion
            #[ensures(result == Self::null_logic())]
```

Otherwise, `null_logic` should be removed.

Reply:
this is not sound wrt the existence of multiple nulls. We could remove `Self::null_logic()` if it's too tempting to incorrectly use to test nullity (instead of `.is_null_logic()`). In principle, `null_logic()` gives you the ability to talk about "a null pointer" in the logic, but I don't know how useful that is...

Reply:
What is not sound is to assume that all the null types are equal. This is not what I am proposing here.

What I am proposing is essentially to assume that the `null` function is deterministic (it always returns the same value). I am not completely sure, but this seems very reasonable.

Now, you may argue that this is useless, which I may agree on. But if we do not add this post-condition, then we should remove `null_logic`, because I think this is useless too.

Reply:
Indeed, I agree. To sum things up, there are two choices:
- keep `Self::null_logic()` and specify that it is equal to the result of the `null()` program function (thus specifying that `null()` is deterministic;
- remove `Self::null_logic` and only use `is_null()` / compare `addr_logic()` to 0. 

I'm leaning towards option 2, because having a logical NULL is not so useful and gives the wrong idea (that there is only one NULL pointer). Using `is_null` instead of comparing to NULL is the better option.

Reply:
Makes sense, and this does not prevent us from adding this in the future.

Reply:
I pushed changes implementing Option 2, with an additional helper function `is_null_logic()` which simply unfolds to `.addr_logic() == 0`. (I found this convenient in practice because it mirrors `is_null()` in program code.)